server/services/v1/cache/sessions.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2022-2023 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rs/zerolog/log"
+	"github.com/tigrisdata/tigris/errors"
+	"github.com/tigrisdata/tigris/server/metadata"
+	"github.com/tigrisdata/tigris/server/request"
+	"github.com/tigrisdata/tigris/server/transaction"
+)
+
+type Session interface {
+	Execute(ctx context.Context, runner QueryRunner) (*CacheResponse, error)
+	TxExecute(ctx context.Context, runner TxQueryRunner) (*CacheResponse, error)
+}
+
+type SessionManager struct {
+	sync.RWMutex
+
+	txMgr         *transaction.Manager
+	tenantMgr     *metadata.TenantManager
+	tenantTracker *metadata.CacheTracker
+	versionH      *metadata.VersionHandler
+}
+
+func NewSessionManager(txMgr *transaction.Manager, tenantMgr *metadata.TenantManager, versionH *metadata.VersionHandler, tenantTracker *metadata.CacheTracker) *SessionManager {
+	return &SessionManager{
+		txMgr:         txMgr,
+		tenantMgr:     tenantMgr,
+		versionH:      versionH,
+		tenantTracker: tenantTracker,
+	}
+}
+
+func (sessMgr *SessionManager) Execute(ctx context.Context, runner QueryRunner) (*CacheResponse, error) {
+	namespaceForThisSession, err := request.GetNamespace(ctx)
+	if err != nil {
+		return nil, err
+	}
+	tenant, err := sessMgr.tenantMgr.GetTenant(ctx, namespaceForThisSession)
+	if err != nil {
+		log.Warn().Err(err).Msgf("Could not find tenant, this must not happen with right authn/authz configured")
+		return nil, errors.NotFound("Tenant %s not found", namespaceForThisSession)
+	}
+
+	// check and reload caches
+	if _, err = sessMgr.tenantTracker.InstantTracking(ctx, nil, tenant); err != nil {
+		return nil, err
+	}
+	return runner.Run(ctx, tenant)
+}
+
+func (sessMgr *SessionManager) TxExecute(ctx context.Context, runner TxQueryRunner) (*CacheResponse, error) {
+	namespaceForThisSession, err := request.GetNamespace(ctx)
+	if err != nil {
+		return nil, err
+	}
+	tenant, err := sessMgr.tenantMgr.GetTenant(ctx, namespaceForThisSession)
+	if err != nil {
+		log.Warn().Err(err).Msgf("Could not find tenant, this must not happen with right authn/authz configured")
+		return nil, errors.NotFound("Tenant %s not found", namespaceForThisSession)
+	}
+
+	// check and reload caches
+	if _, err = sessMgr.tenantTracker.InstantTracking(ctx, nil, tenant); err != nil {
```

Comment:
add a ToDo here that we need to revisit this as reloading like this will be expensive.

Reply:
done.